### ISSUE TITLE
Memory Leak and bursts of high CPU usage

### ISSUE THREAD

**Issue by Nensec** [NillerMedDild/Enigmatica2#289]
**Issue Description**

I am running a dedicated server for a group of friends, but very quickly as people play the memory usage shoots up insanely. Added with this is a lot of server lag, often in the range of 15-20 seconds. During these lag spikes I can see the memory increasing slowly as the CPU usage is at 30-35% and the game is completely unresponsive. From what I could tell this seem to occur the most as people were exploring. In an attempt to remedy this I use the chunk pregen mod included in the pack to generate 450chunk radius from spawn, so I can safely assume it is not chunk generation.

**Modpack version (Do not use "latest")**

1.57

**Logs**

I had restarted the server since, so the only thing I have is the debug log of the previous restart where it occured. It is too big for gist so here is the gzip.
[debug-1.log.gz](https://github.com/NillerMedDild/Enigmatica2/files/2804175/debug-1.log.gz)
In this playthrough only 2 people played since previous restart and latest restart. I did not go out exploring much so I did not happen to trigger it. My friend (ReaperCaster) however seemed to have caused it to happen a few times according to the log. When I came back to the PC the ram usage was at almost 15gb as shown in the screenshot below.

**Screenshots**

https://imgur.com/1JlJAJh

**Additional context**

The following mods were added server side:
DiscordIntegrations ( Simply relays chat back and forth between Discord and Minecraft )
DynMap ( issue occured before adding DynMap  )

The server has a max of 16gb (out of 32gb available in the machine) assigned.
Server files are stored on an SSD.
CPU is a 3.5ghz i5 6600K

If I notice more correlations I will of course update this issue as right now it is probably not enough to pinpoint what mod might be causing it.

**Comment by NillerMedDild** [comment]
That is quite alarming. As I understand it from the log, the leak appears to happen because of world generation, but once the world generation has occured it keeps leaking.

To be completely honest I do not have the time nor ressources to troubleshoot this issue, but an automatic server restart every 12 hours should fix it, at least for now.

**Comment by Nensec** [comment]
Sadly that isn't really a viable solution since the memory increase is gradual but continuous. My friend was only online for an hour and he bumped up the memory usage from about 1.5gb to 15gb : /

**Comment by NillerMedDild** [comment]
It could be related to https://github.com/NillerMedDild/Enigmatica2Expert/issues/703

**Comment by NillerMedDild** [comment]
Have you been to the Twilight Forest either of you?

**Comment by Nensec** [comment]
Nobody has been there on our server

**Comment by jeffwcollins** [comment]
It’s not from Twilight Forest on mine either, as no one has been there, and the only Dim thats loaded is the overworld when we start noticing lag via an increase in cpu and memory usage.

**Comment by NillerMedDild** [comment]
I found this issue which may be related, https://github.com/Azanor/thaumcraft-beta/issues/1455
Under certain circumstances chunks seem to be generating flux on their own, could you check your servers for odd flux values in chunks that have not been exposed to Thaumcraft dabbling?

**Comment by NillerMedDild** [comment]
Also, how do you launch your servers? Do you use the server files' ServerStart script?

**Comment by NillerMedDild** [comment]
Alright, I have a possible solution. Would you be willing to test it out? 
I changed two Forge configs to:
1) Make sure chunks cannot (should not) be rapidly unloaded-then-loaded
2) Keep a cache of recently loaded chunks

You will need to change `config/forge.cfg` line 80 to `I:dimensionUnloadQueueDelay=100` and `config/forgeChunkLoading.cfg` line 26 to `I:dormantChunkCacheSize=64`

**Comment by Nensec** [comment]
I use the serverstart script yes (which btw is amazing!)
In half an hour I'll have time to test it out

**Comment by NillerMedDild** [comment]
I know, it has saved me so much time :D

Awesome! I'm going to sleep in a few minutes, but I will hopefully have some time to troubleshoot further tomorrow. (10 hours from now)

**Comment by Nensec** [comment]
Well I can safely say this did not fix it.
I started the server and logged in: https://i.imgur.com/X1uru9u.png
I went exploring and suddenly everything went stuttering, the server was peaking out at like 75% cpu usage: https://i.imgur.com/mB4Uhd7.png
This was 5 minutes after logging in and exploring: https://i.imgur.com/pYBQYc7.png https://i.imgur.com/ButYfJ9.png

Here are the logs
[debug.log](https://github.com/NillerMedDild/Enigmatica2/files/2805133/debug.log)
[latest.log](https://github.com/NillerMedDild/Enigmatica2/files/2805134/latest.log)

**Comment by NillerMedDild** [comment]
That looks like fairly normal world generation lag. Are you sure you pregenned that area?

**Comment by Nensec** [comment]
I was moving on foot so not fast at all. I was at about 4000 blocks out and I pregenned 450 chunks (~7200 blocks) radius

**Comment by NillerMedDild** [comment]
Damnit. You mentioned you allocated 16GB to the server, could you try with less please? 8GB for example.
Really going to bed now :P gn.

**Comment by Nensec** [comment]
Big update:

We tried to do some testing but honestly we couldn't get very far. My friend was exploring only briefly before I saw my CPU spike up to 80% usage on the server process and the memory almost instantly went up to 8.9gb (out of 8gb max assigned..?). The server at that point simply stopped responding completely resulting in us eventually disconnecting on the client side with a time out.

The server is completely unresponsive, it doesn't even take console input anymore.

I could give the logs but they simply stopped too, nothing interesting in there

**Comment by NillerMedDild** [comment]
Damnit.

**Comment by Nensec** [comment]
Ok so interesting update:

I've reverted back the chunk settings and gone back to 16gb of ram.

We had a 3rd person join who really didn't move anywhere much.


We went exploring and it kind of did the exact same thing but the server did not stall this time.
After about half an hour of constant 15-28sec lag spikes it just kinda stopped. It's been 20 minutes now and the are no lagspikes or cpu spikes. RAM usage is stuck at cap though at 15,780MB usage atm.

Here are the logs:
[debug - Copy.log](https://github.com/NillerMedDild/Enigmatica2/files/2808476/debug.-.Copy.log)
[latest - Copy.log](https://github.com/NillerMedDild/Enigmatica2/files/2808477/latest.-.Copy.log)

The server went up to max ram usage pretty much at the first to second huge lag spike

**Comment by NillerMedDild** [comment]
I guess that's good? :P

**Comment by Nensec** [comment]
I suppose that depends on how good this is with 0 players online :P
![image](https://user-images.githubusercontent.com/672037/51930799-aa27a880-23fb-11e9-80d6-434e712b00b3.png)

**Comment by NillerMedDild** [comment]
Yikes. I think it's time to try some new java arguments. Try adding these to the end of line 6 of `settings.cfg`, but just before the `;`

`-XX:+UseG1GC -Dsun.rmi.dgc.server.gcInterval=2147483646 -XX:G1NewSizePercent=20 -XX:G1ReservePercent=20 -XX:MaxGCPauseMillis=50 -XX:G1HeapRegionSize=32M`

**Comment by Nensec** [comment]
Issues persist sadly, so gonna update to 1.58 and worst case create a new world for now

**Comment by Nensec** [comment]
1.58 seems stable. Will keep an eye on it and if it stays I guess somewhere along the line of updating to 1.58 our issue was fixed

**Comment by Nensec** [comment]
May have found the culprit, or one of em. I encountered a huge cpu spike and memory usage went from 5 to 11 with this in the console log:
```[00:36:44] [Server thread/INFO] [Astral Sorcery]: Load CachedWorldData 'structures' for world 0
[00:36:45] [Server thread/INFO] [Astral Sorcery]: Loading of 'structures' for world 0 finished.
[00:36:45] [Server thread/WARN] [minecraft/MinecraftServer]: Can't keep up! Did the system time change, or is the server overloaded? Running 16293ms behind, skipping 325 tick(s)
[00:37:41] [Server thread/WARN] [minecraft/MinecraftServer]: Can't keep up! Did the system time change, or is the server overloaded? Running 17440ms behind, skipping 348 tick(s)```

**Comment by NillerMedDild** [comment]
Would you mind reporting that to Astral Sorcery? :)

**Comment by Nensec** [comment]
The entire day the server has been running fine. During my own play time I've encountered a couple of 20-30 sec lagspikes but since they occured as I was stripmining and being alone on the server it was hard to determine what caused them.

Overall it seems that the cause of the constant lag and high CPU usage has disappeared with updating to 1.58, though some notable moments still exist. The only correlation I've noticed so far is the one above where Astral Sorcery seems to load in every structure it ever created in the overworld for some reason into a cache of 6gb only for that memory usage to vanish at a later point in time.

I'd say that this issue is "fixed" unless you want me to keep posting updates on when these lagspikes happen? Maybe if @jeffwcollins also updates to 1.58 we can see if their server also runs smoother now.

**Comment by NillerMedDild** [comment]
Yep, unless Jeffwcollins has something to add I'd call this closed :)
